content/blog/2016/2016-12-13-jenkins-puppet-enterprise-plugin.adoc

```diff
@@ -0,0 +1,271 @@
+following
+link:https://docs.puppet.com/pe/latest/rbac_token_auth.html#generating-a-token-for-use-by-a-service[the
+instructions on the docs site.] Next, create a new Jenkins Credentials item
+with Kind *Secret text* and the *Secret* value the Puppet Enterprise RBAC
+token. It's highly recommended to give the credential an ID value that's
+descriptive and identifiable. You'll use it in your Pipeline scripts.
+
+In your Jenkinsfile, use the puppet.credentials method to set all future Puppet
+methods to use the RBAC token. For example:
+
+----
+puppet.credentials 'pe-team-token'
+----
+
+=== Hiera Backend
+
+The plugin exposes an HTTP API for performing Hiera data lookups for key/value
+pairs managed by Pipeline jobs. To configure Hiera on the Puppet compile
+master(s) to query the Jenkins Hiera data store backend, use the
+link:https://github.com/crayfishx/hiera-http[hiera-http] backend.  On the
+Puppet Enterprise compile master(s), run the following commands:
+
+----
+/opt/puppetlabs/puppet/bin/gem install hiera-http
+/opt/puppetlabs/bin/puppetserver gem install hiera-http
+----
+
+Now you can configure the /etc/puppetlabs/puppet/hiera.yaml file. The following
+configuration instructs Hiera to first look to the Hiera yaml files in the
+Puppet code's environment, then fall back to the http backend. The http backend
+will first query the Hiera data store API looking for the key in the scope with
+the same name as the node. If nothing's found, look for the key in the node's
+environments. You can use any Facter fact to match scope names.
+
+----
+:backends:
+  - yaml
+  - http
+
+:http:
+  :host: jenkins.example.com
+  :port: 8080
+  :output: json
+  :use_auth: true
+  :auth_user: <user>
+  :auth_pass: <pass>
+  :cache_timeout: 10
+  :failure: graceful
+  :paths:
+    - /hiera/lookup?path=%{clientcert}&key=%{key}
+    - /hiera/lookup?path=%{environment}&key=%{key}
+----
+
+Finally, restart the pe-puppetserver process to pick up the new configs.
+----
+/opt/puppetlabs/bin/puppet resource service pe-puppetserver ensure=stopped
+/opt/puppetlabs/bin/puppet resource service pe-puppetserver ensure=running
+----
+
+==== Hiera HTTP Authentication
+
+If Jenkins' Global Security is configured to allow unauthenticated read-only
+access, the 'use_auth', 'auth_pass', and 'auth_user' parameters are
+unnecessary. Otherwise, create a local Jenkins user that has permissions to
+view the Hiera Data Lookup page and use that user's credentials for the
+hiera.yaml configuration.
+
+== Querying the infrastructure
+
+PuppetDB is an extensive data store that holds every bit of information Puppet
+generates and collects across every system Puppet is installed on. PuppetDB's
+provides a sweet query language called
+link:https://docs.puppet.com/puppetdb/4.3/api/query/v4/pql.html[PQL.] With PQL,
+you can ask complex questions of your infrastructure such as "How production
+Red Hat systems are there with the openssl package installed?" or "What are are
```

Comment:
What are are the us-west-2c nodes with the MyApp role that were created in the last 24 hours? ->
What us-west-2c nodes with the MyApp role were created in the last 24 hours?